server/coordinator/scheduler/scheduler_manager.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+package scheduler
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/CeresDB/ceresmeta/pkg/log"
+	"github.com/CeresDB/ceresmeta/server/cluster"
+	"github.com/CeresDB/ceresmeta/server/coordinator"
+	"github.com/CeresDB/ceresmeta/server/coordinator/procedure"
+	"go.uber.org/zap"
+)
+
+const (
+	schedulerInterval = time.Second * 5
+)
+
+// Manager used to manage schedulers, it will register all schedulers when it starts.
+//
+// Each registered scheduler will generate procedures if the cluster topology matches the scheduling condition.
+type Manager interface {
+	ListScheduler() []Scheduler
+
+	Start(ctx context.Context) error
+
+	Stop(ctx context.Context) error
+
+	// Scheduler will be called when received new heartbeat, every scheduler registered in schedulerManager will be called to generate procedures.
+	// Scheduler cloud be schedule with fix time interval or heartbeat.
+	Scheduler(ctx context.Context, topology cluster.Topology) []procedure.Procedure
+}
+
+type ManagerImpl struct {
+	procedureManager procedure.Manager
+	clusterManager   cluster.Manager
+	factory          *coordinator.Factory
+	nodePicker       coordinator.NodePicker
+
+	// This lock is used to protect the following field.
+	lock               sync.RWMutex
+	registerSchedulers []Scheduler
+	isRunning          bool
+	cancel             context.CancelFunc
+}
+
+func NewManager(procedureManager procedure.Manager, clusterManager cluster.Manager, factory *coordinator.Factory) Manager {
+	return &ManagerImpl{
+		procedureManager:   procedureManager,
+		clusterManager:     clusterManager,
+		registerSchedulers: []Scheduler{},
+		isRunning:          false,
+		factory:            factory,
+		nodePicker:         coordinator.NewRandomNodePicker(),
+	}
+}
+
+func (m *ManagerImpl) Stop(_ context.Context) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if m.isRunning {
+		m.cancel()
+		m.registerSchedulers = []Scheduler{}
```

Comment:
Why do we have to clear the registered schedulers?

Reply:
`schedulerManager` will start after elected and stop before transfer,  if we do not clear the registered schedulers, it will recall `initRegister` in start when this server is elected to leader again, which may be add duplicate schedulers.

Reply:
If so, I guess it is better to clear the slice by:
```suggestion
		m.registerSchedulers = m.registerSchedulers[:0]
```